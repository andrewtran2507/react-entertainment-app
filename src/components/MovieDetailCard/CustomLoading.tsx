import React, { FC } from 'react';
import { Col, Row, Skeleton } from 'antd';

import './index.scss';

const CustomLoading: FC = () => {
  return (
    <Row gutter={[16, 16]}>
      <Col span={3}></Col>
      <Col span={18}>
        <Skeleton avatar active />
        <Skeleton avatar active />
      </Col>
      <Col span={3}></Col>

      <Col span={3}></Col>
      <Col span={3}>
        <Skeleton active />
        <Skeleton active />
      </Col>
      <Col span={3}>
        <Skeleton active />
        <Skeleton active />
      </Col>
      <Col span={3}>
        <Skeleton active />
      </Col>
      <Col span={3}>
        <Skeleton active />
        <Skeleton active />
      </Col>
      <Col span={3}>
        <Skeleton active />
        <Skeleton active />
      </Col>
      <Col span={3}>
        <Skeleton active />
      </Col>
      <Col span={3}></Col>

      <Col span={3}></Col>
      <Col span={18}>
        <Skeleton active />
        <Skeleton active />
      </Col>
      <Col span={3}></Col>
    </Row>

    // <div className="skeleton-movie-detail-card">
    //   <div className="skeleton-img">
    //     <Skeleton.Image active />
    //   </div>
    //   <Skeleton avatar active />
    //   <Skeleton avatar active />
    //   <br />
    //   <br />
    //   <Skeleton active />
    // </div>
  );
};

export default CustomLoading;
