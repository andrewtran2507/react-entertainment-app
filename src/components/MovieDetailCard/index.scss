.sbdocs-content {
  .movie-detail-card {
    height: 460px;
  }
}
@mixin normal-color-1 {
  color: #FFF;
  text-shadow: #3F51B5 1px 0 5px;
}

@mixin normal-color-2 {
  text-shadow: #eaa6bd 1px 0 5px;
  color: #fff;
}

.movie-detail-body {
  .ant-card-cover {
    position: absolute;
    >img {
      opacity: 0.7;
    }
  }
  .ant-card-body {
    position: relative;
    z-index: 100;
    .content {
      border-bottom: 1px solid var(--primaryColor);
      background-position: left calc((53vw - 170px) - 470px) top;
      background-size: cover;
      .mask {
        background: linear-gradient(to right, rgb(33 6 6) calc(42vw - 170px - 340px), rgb(225 246 10 / 22%) 50%, rgb(53 18 11 / 84%) 100%);
        .ant-image {
          border-width: 0;
          overflow: hidden;
          border-radius: 8px;
          margin: 10px;
        }
        .detail-group {
          padding: 10px;
          width: 100%;
          .movie-title {
            @include normal-color-1;
            padding: 0;
            margin: 0;
            font-size: 16px;
          }
          .main {
            font-size: 24px;
          }
          .info {
            margin: 5px 0;
            .ant-descriptions {
              width: 650px;
            }
            .ant-descriptions-item-label {
              font-weight: 600;
              @include normal-color-2;
            }
            .ant-descriptions-item-content {
              @include normal-color-1;
            }
            .ant-descriptions-row > td {
              padding-bottom: unset;
            }

          }
          .text-header {
            font-weight: 600;
          }
        }
      }
    }
  }
}
.cast-and-producer {
  padding: 20px;
  .ant-descriptions-item-label {
    font-weight: 600;
    color: black;
  }
  .ant-descriptions-item-content {
    display: unset !important;
  }
  .ant-card-meta-description {
    font-size: 10px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  .ant-card-cover {
    img {
      width: inherit;
    }
  }
}
