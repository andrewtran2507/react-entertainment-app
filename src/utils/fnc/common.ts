import { IErrorResponse, TErrors, defElm } from 'utils/constant';

export const getImageULR = (
  path: string | null | undefined,
  folder: string,
): string =>
  path
    ? `${process.env.REACT_APP_IMAGE_URL}/${folder}${path}`
    : '/assets/images/no-img.png';

export const mockMatchMedia = (): void => {
  window.matchMedia = (query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  });
};

export const getErrorResponse = (error: TErrors): IErrorResponse => {
  let objError: IErrorResponse = {
    status_code: 2,
    status_message: 'Something when wrong!',
    success: false,
  };
  if (!error) {
    return objError;
  }
  if (error.response.data) {
    objError = error?.response?.data;
  }
  if (error?.name === 'ZodError') {
    objError.status_message =
      error?.errors
        ?.map(
          (d: { path: string[]; message: string }) =>
            `${d?.path[d?.path.length - 1]} ${d?.message}`,
        )
        .join(', ') || objError.status_message;
  }
  return objError;
};

export const sleep = (ms: number) =>
  new Promise((resolve) => setTimeout(resolve, ms));

export const elmQuerySelector = (
  elm: Element,
  queryStr: string,
  isAll = false,
) =>
  (isAll
    ? Array.from(elm.querySelectorAll(queryStr))
    : elm.querySelector(queryStr)) || defElm;
