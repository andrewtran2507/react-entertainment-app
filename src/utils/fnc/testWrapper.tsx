import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import { movieReducer } from 'pages/Home/Redux';
import { movieDetailReducer } from 'pages/MovieDetail/Redux';
import MainLayout from 'components/Layout/MainLayout';
import { IMovieState } from 'pages/Home/Redux/types';
import { IMovieDetailState } from 'pages/MovieDetail/Redux/types';
import { ToolkitStore } from '@reduxjs/toolkit/dist/configureStore';
const defStore = {
  reducer: { movie: movieReducer, movieDetail: movieDetailReducer },
};

function render(
  ui: ReactElement,
  {
    preloadedState = {},
    store = configureStore({ ...defStore, preloadedState }),
    ...renderOptions
  }: {
    preloadedState?: object;
    store?: ToolkitStore<{
      movie: IMovieState;
      movieDetail: IMovieDetailState;
    }>;
  },
) {
  function Wrapper({ children }: { children: ReactElement }) {
    return (
      <Provider store={store}>
        <BrowserRouter>
          <MainLayout>{children}</MainLayout>
        </BrowserRouter>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
