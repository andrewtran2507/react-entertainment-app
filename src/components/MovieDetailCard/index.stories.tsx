import { Meta, StoryObj } from '@storybook/react';
import {
  reactRouterParameters,
  withRouter,
} from 'storybook-addon-react-router-v6';
import MovieDetailCard from './index';
import movieDetailItem from './mock';

const meta: Meta<typeof MovieDetailCard> = {
  title: 'components/MovieDetailCard',
  component: MovieDetailCard,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
    reactRouter: reactRouterParameters({
      location: {
        pathParams: { movieId: 385687 },
      },
    }),
  },
  tags: ['autodocs'],
  argTypes: {
    loading: { control: 'boolean' },
    movieDetailItem: { control: 'object' },
  },
};

export default meta;

type Story = StoryObj<typeof MovieDetailCard>;

export const OnLoading: Story = {
  args: {
    loading: true,
    movieDetailItem: null,
  },
};

export const UnLoading: Story = {
  args: {
    loading: false,
    movieDetailItem,
  },
};
