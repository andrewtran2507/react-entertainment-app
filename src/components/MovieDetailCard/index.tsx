import React, { FC, useMemo, memo } from 'react';
import { Card, Flex, Typography, Image, Descriptions, Row, Col } from 'antd';
import { LazyLoadImage } from 'react-lazy-load-image-component';
import { getImageULR } from 'utils/fnc/common';
import { imageFolder, movieInfo, moviePlusInfo } from 'utils/constant';
import CustomLoading from './CustomLoading';
import {
  ZMovieDetailModel,
  ZExtendItemModel,
} from 'pages/MovieDetail/Redux/types';

import './index.scss';
const { Meta } = Card;

const CardRenderer: FC<{ item: ZExtendItemModel }> = ({ item }) => (
  <Card
    hoverable
    cover={
      <LazyLoadImage
        alt="profile image"
        effect="blur"
        wrapperProps={{
          style: { transitionDelay: '1s' },
        }}
        src={getImageULR(item?.profile_path, imageFolder['220x330'])}
      />
    }
  >
    <Meta title={item?.name} description={item?.job || item?.character} />
  </Card>
);

const CreditsRenderer: FC<{ data: ZExtendItemModel[] }> = ({ data }) => (
  <Row gutter={[8, 16]}>
    {data?.map((item: ZExtendItemModel) => (
      <Col
        xxl={{ span: 3 }}
        xl={{ span: 4 }}
        md={{ span: 6 }}
        sm={{ span: 8 }}
        xs={{ span: 8 }}
        key={item?.id} // EX: id = 609682 not an index number
      >
        <CardRenderer item={item} />
      </Col>
    ))}
  </Row>
);

const MovieDetailCard: FC<{
  loading: boolean;
  movieDetailItem: ZMovieDetailModel | null;
}> = ({ loading = true, movieDetailItem }) => {
  const covertImage = useMemo(
    () => getImageULR(movieDetailItem?.backdrop_path, imageFolder['1920x800']),
    [movieDetailItem?.backdrop_path],
  );

  const posterImage = useMemo(
    () => getImageULR(movieDetailItem?.poster_path, imageFolder['300x450']),
    [movieDetailItem?.poster_path],
  );

  const descriptionsMovieData = useMemo(
    () =>
      movieInfo?.map((data) => ({
        ...data,
        children:
          data.key === 'vote_average'
            ? (movieDetailItem?.vote_average || 0) * 10
            : data.key === 'production_countries'
            ? movieDetailItem?.production_countries[0]?.iso_3166_1 || 'N/A'
            : data.key === 'genres'
            ? movieDetailItem?.genres?.map((item) => item?.name).join(', ') ||
              'N/A'
            : `${movieDetailItem?.budget?.toLocaleString() || 0} USD`,
      })),
    [movieInfo, movieDetailItem],
  );

  const descriptionsMoviePlusData = useMemo(
    () =>
      moviePlusInfo?.map((data) => ({
        ...data,
        label: <Typography.Title level={5}>{data.label}</Typography.Title>,
        children: (
          <CreditsRenderer
            data={
              data.key === 'cast'
                ? movieDetailItem?.credits?.cast?.filter(
                    (item) => item?.popularity > 35,
                  ) || []
                : movieDetailItem?.credits?.crew?.filter(
                    (item) => item?.job?.includes('Producer'),
                  ) || []
            }
          />
        ),
      })),
    [moviePlusInfo, movieDetailItem],
  );

  return loading ? (
    <CustomLoading />
  ) : (
    <div className="movie-detail-card white-color">
      <Card
        className="movie-detail-body"
        bodyStyle={{ padding: 0, overflow: 'hidden' }}
      >
        <div
          className="content"
          style={{
            backgroundImage: `url('${covertImage}')`,
          }}
        >
          <Flex className="mask" justify="center">
            <Image alt="poster image" src={posterImage} />
            <Flex
              className="detail-group"
              vertical
              align="flex-start"
              justify="left"
            >
              <Typography.Title level={2} className="movie-title main">
                {movieDetailItem?.title}
              </Typography.Title>
              <div className="info">
                <Descriptions
                  column={2}
                  layout="horizontal"
                  items={descriptionsMovieData}
                />
              </div>
              <Typography.Text className="movie-title">
                <span className="text-header">Overview</span>
                <br />
                {movieDetailItem?.overview}
              </Typography.Text>
            </Flex>
          </Flex>
        </div>
      </Card>
      <div className="cast-and-producer">
        <Descriptions
          column={1}
          layout="vertical"
          items={descriptionsMoviePlusData}
        />
      </div>
    </div>
  );
};

export default memo(MovieDetailCard);
