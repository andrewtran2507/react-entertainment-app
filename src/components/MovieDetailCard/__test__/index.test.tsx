import React from 'react';
import { RenderResult, act, cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import movieDetailItem from '../mock';
import MovieDetailCard from 'components/MovieDetailCard';
import { mockMatchMedia } from 'utils/fnc/common';

describe('Test Movie Detail Card Component', () => {
  let component: RenderResult | undefined = undefined;
  beforeAll(() => {
    mockMatchMedia();
  });

  afterEach(() => {
    cleanup();
  });

  it('Test to match snapshot of component', async () => {
    await act(() => {
      component = render(
        <MovieDetailCard loading={true} movieDetailItem={movieDetailItem} />,
      );
    });
    const { asFragment } = component || { asFragment: () => DocumentFragment };
    expect(asFragment()).toMatchSnapshot();
  });

  it('Test content on Movie Detail Component', () => {
    const { getByText } = render(
      <MovieDetailCard loading={false} movieDetailItem={movieDetailItem} />,
    );
    expect(getByText('The Marvels')).toBeTruthy();
  });
});
