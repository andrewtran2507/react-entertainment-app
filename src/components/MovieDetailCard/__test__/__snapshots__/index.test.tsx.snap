// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Movie Detail Card Component Test to match snapshot of component 1`] = `
<DocumentFragment>
  <div
    class="ant-row css-dev-only-do-not-override-2i2tap"
    style="margin-left: -8px; margin-right: -8px; row-gap: 16px;"
  >
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
    <div
      class="ant-col ant-col-18 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-with-avatar ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-header"
        >
          <span
            class="ant-skeleton-avatar ant-skeleton-avatar-lg ant-skeleton-avatar-circle"
          />
        </div>
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 50%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-with-avatar ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-header"
        >
          <span
            class="ant-skeleton-avatar ant-skeleton-avatar-lg ant-skeleton-avatar-circle"
          />
        </div>
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 50%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
    <div
      class="ant-col ant-col-18 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
      <div
        class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-3 css-dev-only-do-not-override-2i2tap"
      style="padding-left: 8px; padding-right: 8px;"
    />
  </div>
</DocumentFragment>
`;
